File: backend/src/main/java/com/assignmentpoint/service/MessageService.java
package com.assignmentpoint.service;

import com.assignmentpoint.dto.CreateMessageRequest;
import com.assignmentpoint.dto.MessageDTO;
import com.assignmentpoint.entity.Message;
import com.assignmentpoint.entity.Order;
import com.assignmentpoint.entity.User;
import com.assignmentpoint.exception.ResourceNotFoundException;
import com.assignmentpoint.repository.MessageRepository;
import com.assignmentpoint.repository.OrderRepository;
import com.assignmentpoint.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class MessageService {

    @Autowired
    private MessageRepository messageRepository;

    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private UserRepository userRepository;

    @Transactional
    public MessageDTO sendMessage(CreateMessageRequest request) {
        Order order = orderRepository.findById(request.getOrderId())
                .orElseThrow(() -> new ResourceNotFoundException("Order not found"));

        User sender = userRepository.findById(request.getSenderId())
                .orElseThrow(() -> new ResourceNotFoundException("Sender not found"));

        User receiver = request.getReceiverId() != null ? userRepository.findById(request.getReceiverId())
                .orElseThrow(() -> new ResourceNotFoundException("Receiver not found")) : null;

        Message message = Message.builder()
                .order(order)
                .sender(sender)
                .receiver(receiver)
                .messageText(request.getMessageText())
                .isRead(false)
                .build();

        message = messageRepository.save(message);
        return convertToDTO(message);
    }

    public Page<MessageDTO> getOrderMessages(Long orderId, Pageable pageable) {
        return messageRepository.findByOrderIdOrderByCreatedAtDesc(orderId, pageable)
                .map(this::convertToDTO);
    }

    public Page<MessageDTO> getUserMessages(Long userId, Pageable pageable) {
        return messageRepository.findBySenderIdOrReceiverId(userId, userId, pageable)
                .map(this::convertToDTO);
    }

    @Transactional
    public void markMessageAsRead(Long messageId) {
        Message message = messageRepository.findById(messageId)
                .orElseThrow(() -> new ResourceNotFoundException("Message not found"));

        message.setIsRead(true);
        message.setReadAt(LocalDateTime.now());
        messageRepository.save(message);
    }

    private MessageDTO convertToDTO(Message message) {
        return MessageDTO.builder()
                .id(message.getId())
                .orderId(message.getOrder().getId())
                .senderId(message.getSender().getId())
                .senderName(message.getSender().getFirstName() + " " + message.getSender().getLastName())
                .receiverId(message.getReceiver() != null ? message.getReceiver().getId() : null)
                .receiverName(message.getReceiver() != null ?
                        message.getReceiver().getFirstName() + " " + message.getReceiver().getLastName() : null)
                .messageText(message.getMessageText())
                .isRead(message.getIsRead())
                .readAt(message.getReadAt())
                .createdAt(message.getCreatedAt())
                .build();
    }
}

File: frontend/lib/api/message.service.ts
import apiClient, { type ApiResponse, handleApiError } from "../api-client"

export interface Message {
  id: number
  orderId: number
  senderId: number
  receiverId: number
  messageText: string
  isRead: boolean
  createdAt: string
}

export interface SendMessageRequest {
  orderId: number
  receiverId: number
  messageText: string
}

export const messageService = {
  async getOrderMessages(orderId: number): Promise<Message[]> {
    try {
      const response = await apiClient.get<ApiResponse<Message[]>>(`messages/order/${orderId}`)
      return response.data.data
    } catch (error) {
      throw new Error(handleApiError(error))
    }
  },

  async sendMessage(messageData: SendMessageRequest): Promise<Message> {
    try {
      const response = await apiClient.post<ApiResponse<Message>>("messages", messageData)
      return response.data.data
    } catch (error) {
      throw new Error(handleApiError(error))
    }
  },

  async markAsRead(messageId: number): Promise<void> {
    try {
      await apiClient.patch(`/messages/${messageId}/read`)
    } catch (error) {
      throw new Error(handleApiError(error))
    }
  },

  async getUserMessages(userId: number): Promise<Message[]> {
    try {
      const response = await apiClient.get<ApiResponse<Message[]>>(`messages/user/${userId}`)
      return response.data.data
    } catch (error) {
      throw new Error(handleApiError(error))
    }
  },
}
